ambry-api/src/main/java/com/github/ambry/rest/RestUtils.java

```diff
@@ -910,21 +908,16 @@ public static void setUserMetadataHeaders(Map<String, String> userMetadataMap,
       if (!headerName.startsWith(Headers.USER_META_DATA_HEADER_PREFIX)) {
         headerName = Headers.USER_META_DATA_HEADER_PREFIX + headerName;
       }
-      try {
-        restResponseChannel.setHeader(headerName, headerValue);
-      } catch (IllegalArgumentException iae) {
+      if (StandardCharsets.US_ASCII.newEncoder().canEncode(headerValue)) {
```

Comment:
What happens if the value contains non ascii characters and you pass it to setHeader() as before?  Does it not throw IllegalArgumentException?

Reply:
It's hard to say, I would prefer merging @cgtz 's ambry-client change and verify. If it doesn't work (not able to display foreign language characters), we can use this piece of code to explicitly and proactively encode the head value.

Reply:
Before, setHeader() would convert the value into valid ASCII by replacing any non-ascii characters with "?" symbols. If , after this step, this ascii encoded string did not contain any characters prohibited in headers but still valid ascii (like newline), the setHeader() call will succeed. However, if it did contain something like a newline, it would throw an exception.

Reply:
Here is a test case showing how Netty behaves (== how `NettyResponseChannel` behaves):
```java

  @Test
  public void headerBehavior() {
    HttpResponse responseMetadata = new DefaultHttpResponse(HttpVersion.HTTP_1_1, HttpResponseStatus.OK);

    String nonASCIIStr = "å∫ \uD83D\uDE1D\uD83D\uDE31abcd";
    responseMetadata.headers().set("non-ascii", nonASCIIStr);

    try {
      responseMetadata.headers().set("ascii-with-non-header-chars", "word word \n\r next line");
    } catch (Exception exception) {
      exception.printStackTrace();
    }

    System.out.println(responseMetadata.headers());
  }
```

results in 
```
java.lang.IllegalArgumentException: only ' ' and '\t' are allowed after '\n': word word 
 next line
DefaultHttpHeaders[non-ascii: å∫ 😝😱abcd]
```

Reply:
Here is the actual behavior. The conversion from unicode to question mark happens in the HttpResponseEncoder:
```java
  @Test
  public void headerBehavior() {
    HttpResponse response = new DefaultHttpResponse(HttpVersion.HTTP_1_1, HttpResponseStatus.OK);

    String nonASCIIStr = "å∫ \uD83D\uDE1D\uD83D\uDE31abcd";
    response.headers().set("non-ascii", nonASCIIStr);

    try {
      response.headers().set("ascii-with-non-header-chars", "word word \n\r next line");
    } catch (Exception exception) {
      System.out.println("Got exception while setting \n\r in httpheaders:  " + exception);
    }

    System.out.println("HttpHeaders object" + response.headers());

    EmbeddedChannel channel = new EmbeddedChannel(new HttpResponseEncoder());
    response.headers().set(HttpHeaderNames.TRANSFER_ENCODING, HttpHeaderValues.CHUNKED);
    assertTrue(channel.writeOutbound(response));
    channel.flushOutbound();
    System.out.println("Actual HTTP response buffer:");
    ByteBuf buffer = channel.readOutbound();
    System.out.println(buffer.toString(Charset.defaultCharset()));
  }
```

```
Got exception while setting 
 in httpheaders:  java.lang.IllegalArgumentException: only ' ' and '\t' are allowed after '\n': word word 
 next line
HttpHeaders objectDefaultHttpHeaders[non-ascii: å∫ 😝😱abcd]
Actual HTTP response buffer:
HTTP/1.1 200 OK
non-ascii: �? ????abcd
transfer-encoding: chunked
```

Reply:
Good find! Thanks for writing a unit test to narrow down the root cause. So, we can conclude that if head value has non-ascii but no new line character, `setHeader()` doesn't throw exception and non-ascii characters will be silently replaced with question mark in `HttpResponseEncoder`.